### PR TITLE
Adding a new example to purchase an App Service Managed Certificate (Free Certificate) for APEX (aka root) domain

### PR DESCRIPTION
**Pull request by coelho5br**
# PR Checklist

Check these items before submitting a PR... 

[Contribution Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md)

[Best Practice Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md)

- [ X ] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

## Changelog

*
*
*

### CODE REVIEW

**Comment by StephenWeatherford** [comment]
Sorry for all the activity, working on the pipelines.

File: 101-app-service-create-managed-certificate-apex/README.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+4. **Make sure you Web App does not have any Network restriction or redirection (such as http to https or to another url).**
+
+This arm deployment will:
+
+1. Create a Managed Certificate (Free).
+
+The operation is expected to take some time.
+
+
+You can also deploy using the powershell:
```
Comment (bmoore-msft) [inline 101-app-service-create-managed-certificate-apex/README.md:21]: remove this section as it needs to be the same for all samples in the repo

File: 101-app-service-create-managed-certificate-apex/README.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
+# Create a Managed Certificate (Free) for an APEX domain (aka root/naked domain)
+
+[![Deploy To Azure](https://aka.ms/deploytoazurebutton)](https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-quickstart-templates%2Fmaster%2F101-app-service-create-managed-certificate-apex%2Fazuredeploy.json)
+[![Visualize](https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/1-CONTRIBUTION-GUIDE/images/visualizebutton.svg?sanitize=true)](http://armviz.io/#/?load=https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-quickstart-templates%2Fmaster%2F101-app-service-create-managed-certificate-apex%2Fazuredeploy.json)
+
+
+To deploy this template, you need to have the following resources:
+
+1. The App Service Plan (serverFarm) resource name that the Web App is running.
```
Comment (bmoore-msft) [inline 101-app-service-create-managed-certificate-apex/README.md:9]: add a prereq template for this

File: 101-app-service-create-managed-certificate-apex/metadata.json [hunk 3]
```diff
@@ -0,0 +1,14 @@
+
+{
+  "$schema": "https://aka.ms/azure-quickstart-templates-metadata-schema#",
+  "type": "QuickStart",
+   "itemDisplayName":"Create a App Services Managed Certificate",
+   "description":"Create Managed Certificate (Free) APEX (root domain) for WebApp",
+   "summary":"Create Managed Certificate (Free) APEX (root domain) for WebApp",
+   "githubUsername":"coelho5br",
+   "dateUpdated":"2021-05-06",
+   "validationType": "Manual",
```
Comment (bmoore-msft) [inline 101-app-service-create-managed-certificate-apex/metadata.json:10]: I'm not seeing anything that requires manual validation - remove this when you add the prereq template

**Review by bmoore-msft — CHANGES_REQUESTED** [review]

File: 101-app-service-create-managed-certificate-apex/metadata.json (hunk 3 above)
Comment (coelho5br) [reply 101-app-service-create-managed-certificate-apex/metadata.json:10]: @bmoore-msft 
I'll leave as manual for now, due to dependency on the A record to match where the Web App will be created. If the deployment has access to the Azure DNS zone holding the domain we will run the test, I can add it to the prereq. I only need to know the Resource group of the DNS zone

File: 101-app-service-create-managed-certificate-apex/README.md (hunk 2 above)
Comment (coelho5br) [reply 101-app-service-create-managed-certificate-apex/README.md]: I added, but let me know how to add the A record, if we go to manual review

File: 101-app-service-create-managed-certificate-apex/README.md (hunk 1 above)
Comment (coelho5br) [reply 101-app-service-create-managed-certificate-apex/README.md]: removed

File: 101-app-service-create-managed-certificate-apex/prereqs/prereq.azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+      ]
+   },
+       {
+      "apiVersion": "2018-11-01",
+      "name": "[concat(parameters('websiteName'), '/', parameters('domainName'))]",
+      "type": "Microsoft.Web/sites/hostNameBindings",
+      "location": "[resourceGroup().location]",
+      "properties": {
+        "sslState": "SniEnabled",
+        "thumbprint": "[reference(resourceId('Microsoft.Web/certificates', variables('certificateName'))).Thumbprint]"
```
Comment (bmoore-msft) [inline 101-app-service-create-managed-certificate-apex/prereqs/prereq.azuredeploy.json:57]: This resource doesn't exist in the template and as a prereq does not exist at time of deployment...  is this supposed to be in the main template?

File: 101-app-service-create-managed-certificate-apex/prereqs/prereq.azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,80 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        "domainName": {
+        "type": "string",
+        "metadata" : {
```
Comment (bmoore-msft) [inline 101-app-service-create-managed-certificate-apex/prereqs/prereq.azuredeploy.json:7]: format the json, e.g. SHIFT+ALT+F in VS Code

**Review by bmoore-msft — CHANGES_REQUESTED** [review]
Check the pipeline results for other changes that need to be made...

File: 101-app-service-create-managed-certificate-apex/prereqs/prereq.azuredeploy.json (hunk 5 above)
Comment (coelho5br) [reply 101-app-service-create-managed-certificate-apex/prereqs/prereq.azuredeploy.json:7]: done

File: 101-app-service-create-managed-certificate-apex/prereqs/prereq.azuredeploy.json (hunk 4 above)
Comment (coelho5br) [reply 101-app-service-create-managed-certificate-apex/prereqs/prereq.azuredeploy.json]: Fixed this part, but the hostnamebind will need to have a A DNS record pointing from the root domain to the web app just created. If the account has access to the DNS zone, we can automate this part, or validation will need to be make manually

File: quickstarts/microsoft.web/app-service-managed-certificate-apex/prereqs/prereq.azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+      "dependsOn": [
+        "[resourceId('Microsoft.Web/serverfarms', variables('AppServicePlanName'))]"
+      ]
+    },
+    {
+      "apiVersion": "2018-11-01",
+      "name": "[concat(parameters('websiteName'), '/', parameters('domainName'))]",
+      "type": "Microsoft.Web/sites/hostNameBindings",
+      "location": "[resourceGroup().location]",
+      "properties": {
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/app-service-managed-certificate-apex/prereqs/prereq.azuredeploy.json:55]: this needs a dependency on the parent resource

File: quickstarts/microsoft.web/app-service-managed-certificate-apex/prereqs/prereq.azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    },
+    {
+      "apiVersion": "2018-02-01",
+      "name": "[parameters('websiteName')]",
+      "type": "Microsoft.Web/sites",
+      "location": "[resourceGroup().location]",
+      "properties": {
+        "serverFarmId": "[variables('AppServicePlanName')]"
+      },
+      "dependsOn": [
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/app-service-managed-certificate-apex/prereqs/prereq.azuredeploy.json:46]: move this above properties

See: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#sort-order-of-properties for the proper sort order of properties

File: quickstarts/microsoft.web/app-service-managed-certificate-apex/prereqs/prereq.azuredeploy.json [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    "sku": "s1",
+    "workerSize": "0",
+    "AppServicePlanName": "[concat(toLower(parameters('websiteName')), 'ASP')]",
+  },
+  "resources": [
+    {
+      "apiVersion": "2018-02-01",
+      "name": "[variables('AppServicePlanName')]",
+      "type": "Microsoft.Web/serverfarms",
+      "location": "[resourceGroup().location]",
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/app-service-managed-certificate-apex/prereqs/prereq.azuredeploy.json:29]: Add location parameter - see: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#parameters

File: quickstarts/microsoft.web/app-service-managed-certificate-apex/prereqs/prereq.azuredeploy.json [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    },
+    "websiteName": {
+      "type": "string",
+      "metadata": {
+        "description": "Web App name that will host the Custom Domain. This name should be unique in Azure."
+      }
+    }
+  },
+  "variables": {
+    "certificateName": "[toLower(concat(parameters('domainName')))]",
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/app-service-managed-certificate-apex/prereqs/prereq.azuredeploy.json:19]: remove this and just use the parameter

File: quickstarts/microsoft.web/app-service-managed-certificate-apex/azuredeploy.json [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+            }
+        },
+        "AppServicePlanName": {
+            "type": "string",
+            "metadata": {
+                "description": "App Service Plan Name where certificate will be imported to."
+            }
+        },
+        "Location": {
+            "type": "string",
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/app-service-managed-certificate-apex/azuredeploy.json:18]: add the defaultValue for [resourceGroup().location]

File: quickstarts/microsoft.web/app-service-managed-certificate-apex/azuredeploy.parameters.json [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        "SubjectName": {
+            "value": "contoso.com"
+        },
+        "AppServicePlanName": {
+            "value": "ASP-TEST-CERT"
+        },
+        "Location": {
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/app-service-managed-certificate-apex/azuredeploy.parameters.json:11]: remove this and use the defaultValue

File: quickstarts/microsoft.web/app-service-managed-certificate-apex/prereqs/prereq.azuredeploy.json [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+      "properties": {
+        "azureResourceName": "[concat('Microsoft.Web/sites/', parameters('websiteName'))]",
+        "azureResourceType": "Website",
+        "customHostNameDnsRecordType": "A",
+        "hostNameType": "Managed",
+        "siteName": "[parameters('websiteName')]"
+      }
+    }
+  ],
+  "outputs": {
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/app-service-managed-certificate-apex/prereqs/prereq.azuredeploy.json:64]: match the outputs here with the inputs of the template

File: quickstarts/microsoft.web/app-service-managed-certificate-apex/azuredeploy.parameters.json [hunk 13]
```diff
@@ -0,0 +1,15 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        "SubjectName": {
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/app-service-managed-certificate-apex/azuredeploy.parameters.json:5]: use the GET-PREREQ tokens for the param values here

File: quickstarts/microsoft.web/app-service-managed-certificate-apex/azuredeploy.json [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+        }
+    },
+    "resources": [
+        {
+            "apiVersion": "2019-08-01",
+            "location": "[parameters('Location')]",
+            "name": "[concat(parameters('SubjectName'), '-', parameters('AppServicePlanName'), '-', parameters('Location'))]",
+            "type": "Microsoft.Web/certificates", 
+            "properties": {
+                "serverFarmId": "[resourceId('Microsoft.Web/serverfarms/', parameters('AppServicePlanName'))]",
```
Comment (bmoore-msft) [inline quickstarts/microsoft.web/app-service-managed-certificate-apex/azuredeploy.json:31]: the resourceGroup parameter needs to be added here (and as a param to the template)

**Review by bmoore-msft — CHANGES_REQUESTED** [review]

File: quickstarts/microsoft.web/app-service-managed-certificate-apex/prereqs/prereq.azuredeploy.json (hunk 6 above)
Comment (coelho5br) [reply quickstarts/microsoft.web/app-service-managed-certificate-apex/prereqs/prereq.azuredeploy.json:55]: done

File: quickstarts/microsoft.web/app-service-managed-certificate-apex/prereqs/prereq.azuredeploy.json (hunk 8 above)
Comment (coelho5br) [reply quickstarts/microsoft.web/app-service-managed-certificate-apex/prereqs/prereq.azuredeploy.json:29]: done

File: quickstarts/microsoft.web/app-service-managed-certificate-apex/azuredeploy.json (hunk 10 above)
Comment (coelho5br) [reply quickstarts/microsoft.web/app-service-managed-certificate-apex/azuredeploy.json:18]: done

File: quickstarts/microsoft.web/app-service-managed-certificate-apex/azuredeploy.json (hunk 14 above)
Comment (coelho5br) [reply quickstarts/microsoft.web/app-service-managed-certificate-apex/azuredeploy.json:32]: as on the properties we are informing the serverFarmId, the API will take care of the Resource Group as it should match the app service plan

File: quickstarts/microsoft.web/app-service-managed-certificate-apex/azuredeploy.parameters.json (hunk 13 above)
Comment (bmoore-msft) [reply quickstarts/microsoft.web/app-service-managed-certificate-apex/azuredeploy.parameters.json:5]: You need to match the GET-PREREQ tokens for all parameters with the outputs of the prereq templates.  Look at other examples in the repo if you need a guide (or the contribution guide itself).

File: quickstarts/microsoft.web/app-service-managed-certificate-apex/azuredeploy.json (hunk 14 above)
Comment (bmoore-msft) [reply quickstarts/microsoft.web/app-service-managed-certificate-apex/azuredeploy.json:32]: only if they are deployed to the same resourceGroup - which does not have to be the case.

**Review by bmoore-msft — CHANGES_REQUESTED** [review]

File: quickstarts/microsoft.web/app-service-managed-certificate-apex/azuredeploy.parameters.json [hunk 15]
```diff
@@ -3,13 +3,13 @@
...
-            "value": "GET-PREREQ"
+            "value": "GET-PREREQ-domainName"
         },
         "AppServicePlanName": {
-            "value": "ASP-new"
+            "value": "GET-PREREQ-AppServicePlanName"
         },
-        "Location": {
-            "value": "West US"
+        "resourceGroup": {
```
[the remark this answers is not in the document]
Comment (coelho5br) [reply quickstarts/microsoft.web/app-service-managed-certificate-apex/azuredeploy.parameters.json:11]: Location removed

File: quickstarts/microsoft.web/app-service-managed-certificate-apex/azuredeploy.json [hunk 16]
```diff
@@ -20,14 +20,21 @@
             "metadata": {
                 "description": "Location of app service plan."
             }
+        },
+         "resourceGroupName": {
+            "type": "string",
+            "metadata": {
```
[the remark this answers is not in the document]
Comment (coelho5br) [reply quickstarts/microsoft.web/app-service-managed-certificate-apex/azuredeploy.json:26]: added a defaultValud on the resouceGroupName using the location

[the remark this answers is not in the document]
Comment (bmoore-msft) [reply quickstarts/microsoft.web/app-service-managed-certificate-apex/azuredeploy.json:26]: this still needs the defaultValue

**Review by bmoore-msft — CHANGES_REQUESTED** [review]

File: quickstarts/microsoft.web/app-service-managed-certificate-apex/azuredeploy.json (hunk 16 above)
[the remark this answers is not in the document]
Comment (coelho5br) [reply quickstarts/microsoft.web/app-service-managed-certificate-apex/azuredeploy.json:27]: Done

**Pull request merged**